content/posts/2021-06-25-Treasury-Expenditure-Bug.md

```diff
@@ -0,0 +1,167 @@
+	if nbNonEmptyWindows > 0 {
+		avgSpentPriorWindows = spentPriorWindows / nbNonEmptyWindows
+	} else {
+		avgSpentPriorWindows = int64(b.chainParams.TreasuryExpenditureBootstrap)
+	}
+
+	//...
+```
+
+Note the last `if` statement: if there is at least one preceding window in which a treasury spend tx ("tspend") was mined, then the average used to derive the limit is calculated as the total spent across all tspends over the number of non-empty windows. Otherwise, the bootstrap amount is used.
+
+In other words, a single tspend transaction in the recent blockchain history already switches the expenditure check from using the bootstrap amount to using the historical amount.
+
+Therefore, when the test tspend transaction was published and mined in mainnet, spending a total of 0.1 DCR from the new treasury account, the expenditure check code was "locked" into only allowing a 50% increase from this amount.
+
+While unit tests were written to assert the correct behavior of the expenditure check in the upper limit of operation (that is, when a set of transactions is trying to spend more than allowed) no tests are verifying the behavior in the _lower_ limit operation. Manual testing was also not performed for this particular case.
+
+## Recovering from the Bug
+
+The resulting effects from locking the decentralized treasury on such a low level of possible spending are the following:
+
+  - The payment for contractors for the next few months will have to keep coming from the legacy treasury wallet. This doesn't have any practical implications to contractors.
+  - The CMS operators will have slightly more work to do, given they'll have to create the usual payment transactions.
+  - Operation of the treasury in a decentralized fashion will be delayed for a few more months.
+  
+While those are unfortunate effects, they are not critical. In particular, as presented in the previous section, this issue _cannot_ be used to withdraw additional funds from the treasury.
+
+It's also important to highlight that all the other checks on treasury spends are intact. And that the decision to perform a staggered move of the funds and not completely empty out the legacy treasury wallet is also providing us with plenty of room to manuever around this issue.
+
+Finally, we're now faced with a decision regarding how to recover from here on. There are two primary options:
+
+  1. Wait about 4 months (up to block 598188) until the test tspend leaves the recent policy window and the treasury can spend from the account regularly, then resume payments as is.
+  2. Perform some on-chain vote to fix this issue.
+  
+Being _very_ optimistic, any on-chain vote would take at least 3 months to write, deploy and activate. And we already have a possible upcoming consensus change that is currently being voted in Politeia ([Explicit Version Upgrades Consensus Change](https://proposals.decred.org/record/3a98861)) that we want to include in the next upgrade if possible.
```

Comment:
Starts off sounding like we are going with 1.  Might want to be clear that we are going with option 2. 

Vote might be over before post is up so we might have to update this.

Reply:
Tried to balance things a bit in this paragraph

Reply:
Looks good